pilot/pkg/secrets/kube/multicluster.go

```diff
@@ -55,6 +55,10 @@ func NewMulticluster(client kube.Client, localCluster, secretNamespace string, s
 	return m
 }
 
+func (m *Multicluster) HasSynced() bool {
+	return m.secretController.HasSynced()
```

Comment:
If I am reading this correctly, we are checking that the multicluster secrets sync here. But do we also need to check the gateway credential secrets sync? Like on each addMemberCluster, on NewSecretsController we should wait?

Looking at he serviceregistry code I don't see where it does the same either, but I think we do it somewhere?

Reply:
`secretController.HasSynced` also checks that `RunAndWait` finished for all the clients. 

https://github.com/istio/istio/blob/aff317b83cd3687f8c6c42fc498891d004f6c618/pkg/kube/secretcontroller/secretcontroller.go#L104-L105

The order should be:
1. Start multicluster secrets controller
2. Add the local cluster (during this time we should register any informers)
3. begin the hasSynced check (waits for secrets to come in, which adds clients that we also wait on)

Reply:
I am not sure 

`Client.RunAndWait` is the same as checking ifthe controller has synced? I think that could complete before https://github.com/istio/istio/blob/d949b3edbd21cea85ffb86026adbd11698dde549/pilot/pkg/serviceregistry/kube/controller/controller.go#L607 returns true for example (in particular for service registry since it has to run SyncAll)?

Reply:
Yeah, ideally controllers would expose their own `HasSynced` if they need something more than "are informers synced". For the serviceregistries in particular, 

https://github.com/istio/istio/blob/2144a5d39dbd7ea0bb88e9847c43731cb597e323/pilot/pkg/bootstrap/server.go#L826

And the aggregate registry checks that each of the cluster registries synced.

Reply:
The refactor in the description could also have `add/updateMemberCluster` give back some interface that includes `HasSynced`, then secretcontroller can include that in it's logic to avoid spreading out the hassynced/synctimeout logic too much.

Reply:
Ok makes sense, I guess looking at SecretsController more if the informer is synced we are good to go since we just the Lister anyways.

side note - should we rename SecretsController to CredentialController or something? Its really confusing. I named it so I will take the blame...